README.md
# dash-resumable-upload

Dash Resumable Upload component for large files.

This is an alternate (to dcc.Upload) dash upload component which uses [resumable.js](https://github.com/23/resumable.js) rather than just the HTML5 File API.

To use it, you must wrap your `app.server` with the `dash_resumable_upload(app.server, 'path/where/data/goes')` method. This creates a flask endpoint that the uploader sends data to, and puts all uploaded files into this folder.

#### usage.py in action
![](demo.gif)

## Dash

Go to this link to learn about [Dash][].

## Getting started

```sh
# Install dependencies
$ npm install

# Watch source for changes and build to `lib/`
$ npm start
```

## Development

### Demo server

You can start up a demo development server to see a demo of the rendered
components:

```sh
$ builder run demo
$ open http://localhost:9000
```

You have to maintain the list of components in `demo/Demo.react.js`.

### Code quality and tests

#### To run lint and unit tests:

```sh
$ npm test
```

#### To run unit tests and watch for changes:

```sh
$ npm run test-watch
```

#### To debug unit tests in a browser (Chrome):

```sh
$ npm run test-debug
```

1. Wait until Chrome launches.
2. Click the "DEBUG" button in the top right corner.
3. Open up Chrome Devtools (`Cmd+opt+i`).
4. Click the "Sources" tab.
5. Find source files
  - Navigate to `webpack:// -> . -> spec/components` to find your test source files.
  - Navigate to `webpack:// -> [your/repo/path]] -> dash-resumable-upload -> src` to find your component source files.
6. Now you can set breakpoints and reload the page to hit them.
7. The test output is available in the "Console" tab, or in any tab by pressing "Esc".

#### To run a specific test

In your test, append `.only` to a `describe` or `it` statement:

```javascript
describe.only('Foo component', () => {
    // ...
})l
```

### Testing your components in Dash

1. Build development bundle to `lib/` and watch for changes

        # Once this is started, you can just leave it running.
        $ npm start

2. Install module locally (after every change)

        # Generate metadata, and build the JavaScript bundle
        $ npm run install-local

        # Now you're done. For subsequent changes, if you've got `npm start`
        # running in a separate process, it's enough to just do:
        $ python setup.py install

3. Run the dash layout you want to test

        # Import dash-resumable-upload to your layout, then run it:
        $ python my_dash_layout.py


**TODO:** There is a workflow that links your module into `site-packages` which would
make it unnecessary to re-run `2.` on every change: `python setup.py develop`.
Unfortunately, this doesn't seem to work with resources defined in
`package_data`.

See https://github.com/plotly/dash-components-archetype/issues/20


## Installing python package locally

Before publishing to PyPi, you can test installing the module locally:

```sh
# Install in `site-packages` on your machine
$ npm run install-local
```

## Uninstalling python package locally

```sh
$ npm run uninstall-local
```

## Publishing

For now, multiple steps are necessary for publishing to NPM and PyPi,
respectively. **TODO:**
[#5](https://github.com/plotly/dash-components-archetype/issues/5) will roll up
publishing steps into one workflow.

Ask @chriddyp to get NPM / PyPi package publishing accesss.

1. Preparing to publish to NPM

        # Bump the package version
        $ npm version major|minor|patch

        # Push branch and tags to repo
        $ git push --follow-tags

2. Preparing to publish to PyPi

        # Bump the PyPi package to the same version
        $ vi setup.py

        # Commit to github
        $ git add setup.py
        $ git commit -m "Bump pypi package version to vx.x.x"

3. Publish to npm and PyPi

        $ npm run publish-all

## Builder / Archetype

We use [Builder][] to centrally manage build configuration, dependencies, and
scripts.

To see all `builder` scripts available:

```sh
$ builder help
```

See the [dash-components-archetype][] repo for more information.


[Builder]: https://github.com/FormidableLabs/builder
[Dash]: https://github.com/plotly/dash2
[dash-components-archetype]: https://github.com/plotly/dash-components-archetype


CHANGELOG.md
# Change Log for dash-resumable-upload
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

# 0.0.4 - 2019-06-01
### Fixed
- Updated to work with latest version of Dash and React version 16

## 0.0.3 - 2018-06-19
### Fixed
- Fix issue where in some environments, multiple threads will attempt
file reconstruction simultaneously and attempt to read deleted files.

## 0.0.2 - 2018-06-19
### Fixed
- Fix issue where in environments, calling `os.unlink` or `os.remove` to quickly after
closing a file handle results in an error.

## 0.0.1 - 2018-04-07
- Initial release


src/components/__tests__/Upload.test.js
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Upload from '../Upload.react';

describe('Upload', () => {


    it('renders', () => {
        const renderer = new ShallowRenderer();
        renderer.render(<Upload id="test"/>);
        const component = renderer.getRenderOutput();
        expect(component).to.be.ok;
    });
});


usage.py
import dash_resumable_upload
import dash
import dash_html_components as html
from dash.dependencies import Input, Output
import base64

app = dash.Dash('')

dash_resumable_upload.decorate_server(app.server, "uploads")

app.scripts.config.serve_locally = True  # Uploaded to npm, this can work online now too.

app.css.append_css({
    "external_url": "https://codepen.io/rmarren1/pen/eMQKBW.css"
})

app.layout = html.Div([
    dash_resumable_upload.Upload(
        id='upload',
        maxFiles=1,
        maxFileSize=1024*1024*1000,  # 100 MB
        service="/upload_resumable",
        textLabel="Drag and Drop Here to upload!",
        startButton=False
    ),
    html.Div(id='output')
])


def get_img(x):
    return x
    #encode = base64.b64encode(
    #    open("uploads/%s" % (x), 'rb').read()).decode('ascii')
    #return "data:image/jpg;base64,{}".format(encode)


@app.callback(Output('output', 'children'),
              [Input('upload', 'fileNames')])
def display_files(fileNames):
    if fileNames is not None:
        return html.Ul([html.Li(
            html.Img(height="50", width="100", src=get_img(x))) for x in fileNames])
    return html.Ul(html.Li("No Files Uploaded Yet!"))


if __name__ == '__main__':
    app.run_server(debug=True)


dash_resumable_upload/__init__.py
from flask import request as _request
from flask import abort as _abort
import os as _os
import shutil as _shutil
import dash as _dash
import sys as _sys
import time as _time
from .version import __version__

_current_path = _os.path.dirname(_os.path.abspath(__file__))

_components = _dash.development.component_loader.load_components(
    _os.path.join(_current_path, 'metadata.json'),
    'dash_resumable_upload'
)

_this_module = _sys.modules[__name__]


_js_dist = [
    {
        "relative_package_path": "bundle.js",
        "external_url": (
            "https://unpkg.com/dash-resumable-upload@{}"
            "/dash_resumable_upload/bundle.js"
        ).format(__version__),
        "namespace": "dash_resumable_upload"
    }
]

_css_dist = []


for _component in _components:
    setattr(_this_module, _component.__name__, _component)
    setattr(_component, '_js_dist', _js_dist)
    setattr(_component, '_css_dist', _css_dist)


def decorate_server(server, temp_base):
    # resumable.js uses a GET request to check if it uploaded the file already.
    # NOTE: your validation here needs to match whatever you do in the POST
    # (otherwise it will NEVER find the files)
    @server.route("/upload_resumable", methods=['GET'])
    def resumable():
        resumableIdentfier = _request.args.get('resumableIdentifier', type=str)
        resumableFilename = _request.args.get('resumableFilename', type=str)
        resumableChunkNumber = _request.args.get('resumableChunkNumber',
                                                 type=int)

        if not (resumableIdentfier
                and resumableFilename
                and resumableChunkNumber):
            # Parameters are missing or invalid
            _abort(500, 'Parameter error')

        # chunk folder path based on the parameters
        temp_dir = _os.path.join(temp_base, resumableIdentfier)

        # chunk path based on the parameters
        chunk_file = _os.path.join(
            temp_dir,
            get_chunk_name(resumableFilename, resumableChunkNumber)
        )
        server.logger.debug('Getting chunk: %s', chunk_file)

        if _os.path.isfile(chunk_file):
            # Let resumable.js know this chunk already exists
            return 'OK'
        else:
            # Let resumable.js know this chunk does not exists
            # and needs to be uploaded
            _abort(404, 'Not found')

    # if it didn't already upload, resumable.js sends the file here
    @server.route("/upload_resumable", methods=['POST'])
    def resumable_post():
        resumableTotalChunks = _request.form.get('resumableTotalChunks',
                                                 type=int)
        resumableChunkNumber = _request.form.get('resumableChunkNumber',
                                                 default=1, type=int)
        resumableFilename = _request.form.get('resumableFilename',
                                              default='error', type=str)
        resumableIdentfier = _request.form.get('resumableIdentifier',
                                               default='error', type=str)

        # get the chunk data
        chunk_data = _request.files['file']

        # make our temp directory
        temp_dir = _os.path.join(temp_base, resumableIdentfier)
        if not _os.path.isdir(temp_dir):
            _os.makedirs(temp_dir)

        # save the chunk data
        chunk_name = get_chunk_name(resumableFilename, resumableChunkNumber)
        chunk_file = _os.path.join(temp_dir, chunk_name)

        # make a lock file
        lock_file_path = _os.path.join(
            temp_dir,
            '.lock_{:d}'.format(resumableChunkNumber)
        )

        with open(lock_file_path, 'a'):
            _os.utime(lock_file_path, None)
        chunk_data.save(chunk_file)
        server.logger.debug('Saved chunk: %s', chunk_file)
        _os.unlink(lock_file_path)

        # check if the upload is complete
        chunk_paths = [
            _os.path.join(temp_dir, get_chunk_name(resumableFilename, x))
            for x in range(1, resumableTotalChunks+1)
        ]
        upload_complete = all([_os.path.exists(p) for p in chunk_paths])

        # combine all the chunks to create the final file
        if upload_complete:
            # Make sure all files are finished writing
            while any([_os.path.isfile(
                      _os.path.join(temp_dir, '.lock_{:d}'.format(chunk))
                      ) for chunk in range(1, resumableTotalChunks + 1)]):
                    _time.sleep(1)
            # Make sure some other chunk didn't trigger file reconstruction
            target_file_name = _os.path.join(temp_base, resumableFilename)
            if not _os.path.exists(target_file_name):
                with open(target_file_name, "ab") as target_file:
                    for p in chunk_paths:
                        with open(p, 'rb') as stored_chunk_file:
                            target_file.write(stored_chunk_file.read())
                server.logger.debug('File saved to: %s', target_file_name)
                _shutil.rmtree(temp_dir)

        return resumableFilename


def get_chunk_name(uploaded_filename, chunk_number):
    return uploaded_filename + "_part_%03d" % chunk_number


src/components/Upload.react.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Resumablejs from 'resumablejs';

export default class Upload extends Component {
    constructor(props) {
        super(props);
        this.state = {
            progressBar: 0,
            messageStatus: '',
            fileList: {files: []},
            isPaused: false,
            isUploading: false,
            isHovered: false,
            isComplete: false
        };
        this.toggleHovered = this.toggleHovered.bind(this)
        this.cancelUpload = this.cancelUpload.bind(this)
        this.pauseUpload = this.pauseUpload.bind(this)
        this.startUpload = this.startUpload.bind(this)
        this.resumable = null;
    }

    componentDidMount() {

        const ResumableField = new Resumablejs({
            target: this.props.service,
            query: {},
            fileType: this.props.filetypes,
            maxFiles: this.props.maxFiles,
            maxFileSize: this.props.maxFileSize,
            fileTypeErrorCallback: () => {
                this.setState({
                    messageStatus: 'Invalid file type!'
                });
            },
            testMethod: 'post',
            testChunks: false,
            headers: {},
            chunkSize: this.props.chunkSize,
            simultaneousUploads: this.props.simultaneousUploads,
            forceChunkSize: false
        });

        ResumableField.assignBrowse(this.uploader);

        //Enable or Disable DragAnd Drop
        if (this.props.disableDragAndDrop === false) {
            ResumableField.assignDrop(this.dropZone);
        }

        ResumableField.on('fileAdded', (file) => {
            this.setState({
                messageStatus: this.props.fileAddedMessage || ' Starting upload! of ' + file.fileName,
                isComplete: false
            });

             if (typeof this.props.onFileAdded === 'function') {
                 this.props.onFileAdded(file, this.resumable);
             } else {
                 ResumableField.upload();
             }
        });

        ResumableField.on('fileSuccess', (file, fileServer) => {

            if (this.props.fileNameServer) {
                let objectServer = JSON.parse(fileServer);
                file.fileName = objectServer[this.props.fileNameServer];
            } else {
                file.fileName = fileServer;
            }
            let currentFiles = this.state.fileList.files;
            currentFiles.push(file);

            let fileNames = this.props.fileNames
            fileNames.push(file.fileName);

            if (this.props.setProps) {
                this.props.setProps({
                      fileNames: fileNames
                });
            }
            this.setState({
                fileList: {files: currentFiles},
                isComplete: true,
                messageStatus: this.props.completedMessage + file.fileName || fileServer
            }, () => {
                if (typeof this.props.onFileSuccess === 'function') {
                    this.props.onFileSuccess(file, fileServer);
                }
            });
        });

        ResumableField.on('progress', () => {


            this.setState({
                isUploading: ResumableField.isUploading()
            });

            if ((ResumableField.progress() * 100) < 100) {
                this.setState({
                    messageStatus: parseInt(ResumableField.progress() * 100, 10) + '%',
                    progressBar: ResumableField.progress() * 100
                });
            } else {
                setTimeout(() => {
                    this.setState({
                        progressBar: 0
                    })
                }, 1000);
            }

        });

        ResumableField.on('fileError', (file, errorCount) => {
            this.props.onUploadErrorCallback(file, errorCount);
        });

        this.resumable = ResumableField;
    }

    cancelUpload() {
        this.resumable.cancel();

        this.setState({
            fileList: {files: []}
        });
    }

    pauseUpload() {
        if (!this.state.isPaused) {
            this.resumable.pause();
            this.setState({
                isPaused: true,
                isUploading: true
            });
        } else {
            this.resumable.upload();
            this.setState({
                isPaused: false,
                isUploading: true
            });
        }
    }

    startUpload() {
        this.setState({
            isPaused: false
        });
    }

    toggleHovered() {
        this.setState({
            isHovered: !this.state.isHovered
        })
    }

    render() {

        let fileList = null;

        let textLabel = null;
        if (this.props.textLabel) {
            textLabel = this.props.textLabel;
        }

        let startButton = null;
        if (this.props.startButton) {
            if (typeof this.props.startButton ==='string' || typeof this.props.startButton ==='boolean' ) startButton = <label>
                <button
                  disabled={this.state.isUploading}
                  className="resumable-btn-start"
                  onClick={this.startUpload}>{this.props.startButton && 'upload'}
                </button>
            </label>;
            else startButton =this.props.startButton
        }

        let cancelButton = null;
        if (this.props.cancelButton) {
            if (typeof this.props.cancelButton ===  'string' ||
                typeof this.props.cancelButton ===  'boolean') cancelButton = <label>
                <button
                  disabled={!this.state.isUploading}
                  className="resumable-btn-cancel"
                  onClick={this.cancelUpload}>{this.props.cancelButton && 'cancel'}
                </button>
            </label>;
            else cancelButton = this.props.cancelButton
        }

        let pauseButton = null;
        if (this.props.pauseButton) {
            if (typeof this.props.pauseButton ===  'string'
                || typeof this.props.pauseButton ===  'boolean') pauseButton = <label>
                <button
                  disabled={!this.state.isUploading}
                  className="resumable-btn-pause"
                  onClick={this.pauseUpload}>
                  {this.props.pauseButton
                   && (this.state.isPaused ? 'resume' : 'pause')}
                </button>
            </label>;
            else pauseButton = this.props.pauseButton
        }

        let getStyle = () => {
            if (this.state.isComplete) {
                return this.props.completeStyle;
            } else if (this.state.isHovered || this.state.isUploading) {
                return this.props.activeStyle;
            } else {
                return this.props.defaultStyle;
            }
        }

        let getClass = () => {
            if (this.props.disabledInput) {
              return this.props.disableClass;
            } else if (this.state.isHovered) {
              return this.props.hoveredClass;
            } else if (this.state.isUploading) {
              return this.props.uploadingClass;
            } else if (this.state.isComplete) {
              return this.props.completeClass;
            } else if (this.state.isPaused) {
              return this.props.completeClass;
            } else {
              return this.props.className
            }
        }

        return (
            <div id={this.props.id} className={getClass()} ref={node => this.dropZone = node} >
                <label
                       style={getStyle()}
                       onMouseEnter={this.toggleHovered}
                       onMouseLeave={this.toggleHovered}
                       >
                         {this.state.messageStatus == '' ? textLabel : this.state.messageStatus}
                    <input
                        ref={node=> this.uploader = node}
                        type="file"
                        className='btn'
                        name={this.props.id + '-upload'}
                        accept={this.props.fileAccept || '*'}
                        disabled={this.props.disableInput || false}
                        style={{'opacity': '0',
                                'width': '0.1px%',
                                'height': '0.1px%',
                                'position': 'absolute',
                                'overflow': 'hidden',
                                'z-index': '-1'}}
                    />
                </label>
                <div className="progress"
                     style={{
                       display: this.state.progressBar === 0 ? 'none' : 'block'
                     }}>
                    <div className="progress-bar"
                         style={{
                           width: this.state.progressBar + '%',
                           height: '100%'
                         }}></div>
                    </div>
                {fileList}
                {startButton}
                {pauseButton}
                {cancelButton}
            </div>
        );
    }
}

Upload.propTypes = {
    /**
     * Maximum number of files that can be uploaded in one session
     */
    maxFiles: PropTypes.number,

    /**
     * Maximum size per file in bytes.
     */
    maxFileSize: PropTypes.number,

    /**
     * Size of file chunks to send to server.
     */
    chunkSize: PropTypes.number,

    /**
     * Number of simultaneous uploads to select
     */
    simultaneousUploads: PropTypes.number,

    /**
     * The service to send the files to
     */
    service: PropTypes.string,

    /**
     * Class to add to the upload component by default
     */
    className: PropTypes.string,

    /**
     * Class to add to the upload component when it is hovered
     */
    hoveredClass: PropTypes.string,

    /**
     * Class to add to the upload component when it is disabled
     */
    disabledClass: PropTypes.string,

    /**
     * Class to add to the upload component when it is paused
     */
    pausedClass: PropTypes.string,

    /**
     * Class to add to the upload component when it is complete
     */
    completeClass: PropTypes.string,

    /**
     * Class to add to the upload component when it is uploading
     */
    uploadingClass: PropTypes.string,

    /**
     * Style attributes to add to the upload component
     */
    defaultStyle: PropTypes.object,

    /**
     * Style when upload component is hovered over
     */
    activeStyle: PropTypes.object,

   /**
    * Style when upload is completed (upload finished)
    */
    completeStyle: PropTypes.object,

    /**
     * The string to display in the upload component
     */
    textLabel: PropTypes.string,

    /**
     * Message to display when upload completed
     */
    completedMessage: PropTypes.string,

    /**
     * The names of the files uploaded
     */
    fileNames: PropTypes.arrayOf(PropTypes.string),

    /**
     * List of allowed file types, e.g. ['jpg', 'png']
     */
    filetypes: PropTypes.arrayOf(PropTypes.string),

    /**
     * Whether or not to have a start button
     */
    startButton: PropTypes.bool,

    /**
     * Whether or not to have a pause button
     */
    pauseButton: PropTypes.bool,

    /**
     * Whether or not to have a cancel button
     */
    cancelButton: PropTypes.bool,

    /**
     * Whether or not to allow file drag and drop
     */
    disableDragAndDrop: PropTypes.bool,

    /**
     * Dash-supplied function for updating props
     */
    setProps: PropTypes.func,

    /**
     * User supplied id of this component
     */
    id: PropTypes.string

}

Upload.defaultProps = {
    maxFiles: 1,
    maxFileSize: 1024 * 1024 * 10,
    chunkSize: 1024 * 1024,
    simultaneuosUploads: 1,
    service: '/upload',
    className: 'resumable-default',
    hoveredClass: 'resumable-hovered',
    completeClass: 'resumable-complete',
    disabledClass: 'resumable-disabled',
    pausedClass: 'resumable-paused',
    uploadingClass: 'resumable-uploading',
    defaultStyle: {},
    activeStyle: {},
    completeStyle: {},
    textLabel: 'Click Here to Select a File',
    completedMessage: 'Complete! ',
    fileNames: [],
    filetypes: undefined,
    startButton: true,
    pauseButton: true,
    cancelButton: true,
    disableDragAndDrop: false,
    id: 'default-uploader-id'
};
